File: packages/connector/services/google_calendar/sync_events.py
"""Google Calendar イベント同期

日次同期用。指定日数分のイベントを取得してraw層に保存。
2500件を超える場合はページネーションで取得。
終了日から開始日に向かって遡りながら取得。
"""

import time
from datetime import date, timedelta
from typing import Any, TypedDict

from db.raw_client import upsert_raw_batch, RawRecord
from lib.logger import setup_logger
from services.google_calendar.api_client import fetch_events

logger = setup_logger(__name__)

TABLE_NAME = "google_calendar__events"
API_VERSION = "v3"

# Google Calendar API の制限
# - maxResults: 最大2500件/ページ（デフォルト250）
# - ページネーションはAPI側で自動処理
MAX_EVENTS_PER_CHUNK = 2500


class SyncResult(TypedDict):
    """同期結果"""
    success: bool
    count: int
    elapsed_seconds: float


def _to_raw_record(event: dict[str, Any]) -> RawRecord:
    """APIレスポンスをRawRecordに変換

    source_idは {calendarId}:{eventId} 形式。
    """
    calendar_id = event.get("_calendar_id", "primary")
    event_id = event["id"]
    source_id = f"{calendar_id}:{event_id}"

    return RawRecord(
        source_id=source_id,
        data=event,
    )


async def sync_events(
    days: int = 3,
    start_date: str | None = None,
    end_date: str | None = None,
) -> SyncResult:
    """イベントを同期

    終了日から開始日に向かって遡りながら取得。
    2500件を超える場合はページネーションで自動処理。

    Args:
        days: 同期する日数（今日から遡る、start_date/end_date未指定時）
        start_date: 開始日（YYYY-MM-DD）
        end_date: 終了日（YYYY-MM-DD）

    Returns:
        同期結果
    """
    start_time = time.perf_counter()

    # 日付範囲を決定（end_dateは翌日にして当日分を確実に取得）
    if start_date and end_date:
        start_d = date.fromisoformat(start_date)
        end_d = date.fromisoformat(end_date)
    else:
        end_d = date.today() + timedelta(days=1)
        start_d = date.today() - timedelta(days=days - 1)

    logger.info(f"Starting Google Calendar events sync ({start_d} to {end_d})")

    total_count = 0
    current_end = end_d

    # 終了日から開始日に向かって遡りながら取得
    chunk_num = 0
    while current_end > start_d:
        chunk_num += 1
        chunk_start_str = start_d.isoformat()
        chunk_end_str = current_end.isoformat()

        logger.info(f"Fetching chunk {chunk_num}: {chunk_start_str} to {chunk_end_str}")

        try:
            # API からデータ取得（ページネーション対応済み）
            events = await fetch_events(chunk_start_str, chunk_end_str)
            logger.info(f"Fetched {len(events)} events")

            if not events:
                break

            # DB保存
            records = [_to_raw_record(e) for e in events]
            result = await upsert_raw_batch(TABLE_NAME, records, api_version=API_VERSION)
            total_count += result["total"]
            logger.info(f"Saved {result['total']} events to DB (total: {total_count})")

            # 2500件未満なら全件取得完了
            if len(events) < MAX_EVENTS_PER_CHUNK:
                break

            # 次のチャンク: 最も古いイベントの日付より前から取得
            oldest_event = min(events, key=lambda e: e.get("start", {}).get("dateTime") or e.get("start", {}).get("date", ""))
            oldest_start = oldest_event.get("start", {})
            oldest_date_str = oldest_start.get("dateTime") or oldest_start.get("date")
            if oldest_date_str:
                # dateTimeの場合は日付部分だけ取得
                oldest_date = date.fromisoformat(oldest_date_str[:10])
                current_end = oldest_date
            else:
                break

        except Exception as e:
            elapsed = round(time.perf_counter() - start_time, 2)
            # レートリミットの場合は取得済み分を返して終了
            if "429" in str(e) or "rateLimitExceeded" in str(e):
                logger.warning(
                    f"Google Calendar API rate limit. Saved {total_count} events before limit."
                )
                return SyncResult(
                    success=total_count > 0,
                    count=total_count,
                    elapsed_seconds=elapsed,
                )
            logger.error(f"Google Calendar events sync failed after {elapsed}s: {e}")
            raise

    elapsed = round(time.perf_counter() - start_time, 2)
    logger.info(f"Google Calendar events sync completed: {total_count} events in {elapsed}s")

    return SyncResult(
        success=True,
        count=total_count,
        elapsed_seconds=elapsed,
    )

File: packages/connector/src/services/google-calendar/sync-events.ts
/**
 * Google Calendar Events Sync
 *
 * Daily sync - fetches events for specified days and saves to raw layer.
 * Handles pagination for >2500 events.
 */

import { upsertRawBatch, deleteBySourceIds, RawRecord } from "../../db/raw-client.js";
import { setupLogger } from "../../lib/logger.js";
import { fetchEventsBatch } from "./api-client.js";

const logger = setupLogger("gcal-events");

const TABLE_NAME = "google_calendar__events";
const API_VERSION = "v3";

// Types
export interface SyncResult {
  success: boolean;
  count: number;
  deleted: number;
  elapsedSeconds: number;
}

/**
 * Convert API response to RawRecord
 *
 * source_id is {calendarId}:{eventId} format
 */
function toRawRecord(event: Record<string, unknown>): RawRecord {
  const calendarId = (event._calendar_id as string) || "primary";
  const eventId = event.id as string;
  const sourceId = `${calendarId}:${eventId}`;

  return {
    sourceId,
    data: event,
  };
}

/**
 * Format date as YYYY-MM-DD
 */
function formatDate(date: Date): string {
  return date.toISOString().slice(0, 10);
}

/**
 * Sync events using Batch API
 *
 * Fetches events for the specified date range using batch requests.
 *
 * @param days - Number of days to sync
 * @param startDate - Optional start date (YYYY-MM-DD)
 * @param endDate - Optional end date (YYYY-MM-DD)
 * @returns Sync result
 */
export async function syncEvents(
  days: number = 3,
  startDate?: string,
  endDate?: string
): Promise<SyncResult> {
  const startTime = performance.now();

  // Determine date range
  let startD: Date;
  let endD: Date;

  if (startDate && endDate) {
    startD = new Date(startDate);
    endD = new Date(endDate);
  } else {
    endD = new Date();
    endD.setDate(endD.getDate() + 1);
    startD = new Date();
    startD.setDate(startD.getDate() - (days - 1));
  }

  const startStr = formatDate(startD);
  const endStr = formatDate(endD);

  logger.info(
    `Starting Google Calendar events sync (${startStr} to ${endStr})`
  );

  try {
    // Fetch all events using batch API (includes deleted events with showDeleted=true)
    const events = await fetchEventsBatch(startStr, endStr);
    logger.info(`Fetched ${events.length} events (may include duplicates and cancelled)`);

    if (events.length === 0) {
      const elapsed = Math.round((performance.now() - startTime) / 10) / 100;
      return {
        success: true,
        count: 0,
        deleted: 0,
        elapsedSeconds: elapsed,
      };
    }

    // Separate events by status: confirmed (active) vs cancelled (deleted)
    const confirmedEvents: Record<string, unknown>[] = [];
    const cancelledSourceIds = new Set<string>();

    for (const event of events) {
      const status = event.status as string;
      const calendarId = (event._calendar_id as string) || "primary";
      const eventId = event.id as string;
      const sourceId = `${calendarId}:${eventId}`;

      if (status === "cancelled") {
        cancelledSourceIds.add(sourceId);
      } else {
        confirmedEvents.push(event);
      }
    }

    logger.info(`Events: ${confirmedEvents.length} active, ${cancelledSourceIds.size} cancelled`);

    // Delete cancelled events from DB
    let deletedCount = 0;
    if (cancelledSourceIds.size > 0) {
      deletedCount = await deleteBySourceIds(TABLE_NAME, Array.from(cancelledSourceIds));
    }

    // Deduplicate active events by source_id (events spanning multiple days appear multiple times)
    const recordMap = new Map<string, RawRecord>();
    for (const event of confirmedEvents) {
      const record = toRawRecord(event);
      recordMap.set(record.sourceId, record);
    }
    const records = Array.from(recordMap.values());
    logger.info(`Deduplicated to ${records.length} unique active events`);

    // Save active events to DB
    let upsertedCount = 0;
    if (records.length > 0) {
      const result = await upsertRawBatch(TABLE_NAME, records, API_VERSION);
      upsertedCount = result.total;
    }

    const elapsed = Math.round((performance.now() - startTime) / 10) / 100;
    logger.info(
      `Google Calendar events sync completed: ${upsertedCount} upserted, ${deletedCount} deleted in ${elapsed}s`
    );

    return {
      success: true,
      count: upsertedCount,
      deleted: deletedCount,
      elapsedSeconds: elapsed,
    };
  } catch (error) {
    const elapsed = Math.round((performance.now() - startTime) / 10) / 100;

    // On rate limit, return partial result
    if (
      String(error).includes("429") ||
      String(error).includes("rateLimitExceeded")
    ) {
      logger.warn(`Google Calendar API rate limit after ${elapsed}s`);
      return {
        success: false,
        count: 0,
        deleted: 0,
        elapsedSeconds: elapsed,
      };
    }

    logger.error(`Google Calendar events sync failed after ${elapsed}s: ${error}`);
    throw error;
  }
}
